service/resource/endpoint/create.go

```diff
@@ -1,11 +1,133 @@
 package endpoint
 
-import "context"
+import (
+	"context"
+
+	"github.com/giantswarm/microerror"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	apiv1 "k8s.io/client-go/pkg/api/v1"
+)
 
 func (r *Resource) GetCreateState(ctx context.Context, obj, currentState, desiredState interface{}) (interface{}, error) {
-	return nil, nil
+	currentEndpoints, err := toEndpointSlice(currentState)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	desiredEndpoints, err := toEndpointSlice(desiredState)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	var createState = []Endpoint{}
+	for _, desiredEndpoint := range desiredEndpoints {
+		if !containsEndpoint(currentEndpoints, desiredEndpoint) {
+			createState = append(createState, desiredEndpoint)
+		}
+	}
+
+	if len(createState) == 0 {
+		return nil, nil
+	}
+
+	return createState, nil
 }
 
 func (r *Resource) ProcessCreateState(ctx context.Context, obj, createState interface{}) error {
+	endpointsToCreate, err := toEndpointSlice(createState)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+	for _, endpoint := range endpointsToCreate {
+		k8sService, err := r.k8sClient.CoreV1().Services(endpoint.ServiceNamespace).Get(endpoint.ServiceName, metav1.GetOptions{})
+		if err != nil {
+			return microerror.Mask(err)
+		}
+		k8sEndpoint := &apiv1.Endpoints{
+			TypeMeta: metav1.TypeMeta{
+				APIVersion: "v1",
+			},
+			ObjectMeta: metav1.ObjectMeta{
+				Name: endpoint.ServiceName,
+			},
+			Subsets: []apiv1.EndpointSubset{
+				{
+					Addresses: []apiv1.EndpointAddress{
+						{
+							IP: endpoint.IP,
+						},
+					},
+					Ports: serviceToPorts(k8sService),
+				},
+			},
+		}
+		_, err = r.k8sClient.CoreV1().Endpoints(endpoint.ServiceNamespace).Create(k8sEndpoint)
+		if errors.IsAlreadyExists(err) {
+			// In case the endpoint we tried to create does already exist, we only
+			// need to append the IPs we got in podInfos.
+			err = appendIPs(r, endpoint.ServiceNamespace, endpoint.ServiceName, endpoint.IP)
+			if err != nil {
+				return microerror.Mask(err)
+			}
+		} else if err != nil {
+			return microerror.Mask(err)
+		}
+
+	}
+
 	return nil
 }
+
+func appendIPs(r *Resource, serviceNamespace string, serviceName string, IP string) error {
```

Comment:
These functions also already existed in `k8s-endpoint-updater`, I slightly modified them and left them in this file as they are pretty specific to the create operation. Not sure if they should be put elsewhere.

Reply:
When I see all this and the stuff I think all the state computation should be done in `GetCreateState` (or former steps). `ProcessCreateState` should always only take computed state and apply it to the resource(s) being reconciled. Some quotes from a discussion I had with Joe about this topic. 

> Separating the computation of state used to modify resources and the actual process of apply changes to a resource is the whole point of the resource interface. When you mix it up now you can just don't use it. It is like a best practise. You can also write all your code into the same line, but you should not do it though.

> One task is stateless and is not bound to any dependency because it only implements some algorithm that manipulates data in memory. The other task is almost always about dealing with external dependencies. Testing is easier when it is separated. The code is more structured and the separation of concerns is way better when we just separate these things.

Reply:
I get the reasoning, but wasn't quite aware of it earlier. I simplified the `ProcessCreateState` function. It now takes the inputs of `GetCreateState` and doesn't do much of any computation any more.